### PR TITLE
Podcast Player Submission

### PR DESCRIPTION
**Pull request by ramseylove**
Have the main functionality working. Minimal styling but other than that i'm pretty happy with everything. A bit of clean up and refactoring could be done yet. 

I Changed the layout a few times. Decided I wanted the player out of the browsing window, so I put it in the header. Then I didn't know where I wanted to put each episodes details, so I put that in a modal.

Need to start the json-server along with the react-scripts with `npm run dev` otherwise there will not be any shows. Also Episodes are being limited to only 10 right now.

todos:
styling could be made responsive
pagination of episodes.

possible todos:
integrate with live podcast api to get shows
search functionality

Live app: https://ultimate-podcasting.vercel.app/

### CODE REVIEW

File: podcast-player/podcast-player-app/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+## Start React server and json-server
+
+`npm run dev`
+
+## Dummy data
+
+localhost:3500/podcasts
+
+```json
+podcasts: [
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/README.md]: This isn't valid JSON. See how everything is highlighted in red?

File: podcast-player/podcast-player-app/data/episodes.js [hunk 2]
```diff
@@ -0,0 +1,918 @@
+const episodes = {
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/data/episodes.js]: Any reason you hardcoded all of this data? It's already being fetched live on the client.

File: podcast-player/podcast-player-app/data/server.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+var convert = require("xml-js");
+var jsonServer = require("json-server");
+// var episodes = require("./episodes.json");
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/data/server.js]: [Commented code is still code](https://kentcdodds.com/blog/please-dont-commit-commented-out-code), and reduces maintainability.

File: podcast-player/podcast-player-app/src/utils/api-utils.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    };
+  });
+
+  const cleanedXML = {
+    channel: {
+      copyright: parsedXml.rss.channel.copyright,
+      description: removeHtml(parsedXml.rss.channel.description),
+      image: { ...parsedXml.rss.channel.image },
+      episodes: [...cleanedEpisodes],
+      pubDate: parseDate(parsedXml.rss.channel.pubDate),
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/utils/api-utils.js]: I'm getting an error on this line when I try to load `The Apology Line`.

File: podcast-player/podcast-player-app/src/assets/SVG/circle-right.svg [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<svg version="1.1" xmlns="http://www.w3.org/2000/svg" width="32" height="32" viewBox="0 0 32 32">
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/assets/SVG/circle-right.svg]: Where'd you find these assets? Including assets like this is fine, but there's a little more overhead. Namely - Is the license reflected anywhere? A lot of people use `react-icons` which is multiple large collections of icons.

File: podcast-player/podcast-player-app/src/components/episode/EpisodeDetails.jsx [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import React from "react";
+import { ReactComponent as PlayButton } from "../../assets/SVG/play2.svg";
+
+function EpisodeDetails({ episode, setSelectedEpisodePlaying }) {
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/components/episode/EpisodeDetails.jsx]: Dead code

File: podcast-player/podcast-player-app/src/components/episode/EpisodeList.jsx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+import EpisodeModal from "./EpisodeModal";
+
+function EpisodeList({
+  show,
+  selectedEpisodePlaying,
+  setSelectedEpisodePlaying,
+}) {
+  const [episodes, setEpisodes] = useState([]);
+  const [selectedShow, setSelectedShow] = useState({});
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/components/episode/EpisodeList.jsx]: I feel like `null` is a more semantically appropriate default state value here.

File: podcast-player/podcast-player-app/src/components/episode/EpisodeModal.jsx [hunk 8]
```diff
@@ -0,0 +1,66 @@
+import React from "react";
+import ReactDOM from "react-dom";
+
+import { ReactComponent as PlayButton } from "../../assets/SVG/play2.svg";
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/components/episode/EpisodeModal.jsx]: This isn't a button, it's an icon

File: podcast-player/podcast-player-app/src/components/episode/EpisodeModal.jsx [hunk 9]
```diff
@@ -0,0 +1,66 @@
+import React from "react";
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/components/episode/EpisodeModal.jsx]: React tooling doesn't require you to import `React` anymore

File: podcast-player/podcast-player-app/src/components/episode/EpisodePlayer.jsx [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import React from "react";
+
+function EpisodePlayer({ episode }) {
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/components/episode/EpisodePlayer.jsx]: Stupid simple. It works!

File: podcast-player/podcast-player-app/src/components/header.jsx [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import React from "react";
+import EpisodePlayer from "./episode/EpisodePlayer";
+
+function Header({ selectedEpisodePlaying }) {
+  return (
+    <div className="header">
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/components/header.jsx]: You can [use more semantically appropriate HTML elements](https://developer.mozilla.org/en-US/docs/Glossary/Semantics#semantic_elements). For this, I'd use the `<header>` tag.

File: podcast-player/podcast-player-app/src/components/header.jsx [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+      <h3>Ultimate Podcasting</h3>
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/components/header.jsx]: Expanding on HTML semantics, [your header tags should follow a clear h1-h2-h3-... hierarchy](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements). If you want a different style don't use a different tag - style the tag.

File: podcast-player/podcast-player-app/src/components/show/Show.jsx [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import React from "react";
+
+import { ReactComponent as PlayButton } from "../../assets/SVG/circle-right.svg";
+
+function Show({ id, title, feed, selectShow }) {
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/components/show/Show.jsx]: Lots of these props aren't used.

File: podcast-player/podcast-player-app/src/components/show/ShowList.jsx [hunk 14]
```diff
@@ -0,0 +1,21 @@
+import React from "react";
+import Show from "./Show";
+
+function ShowList({ shows, selectShow }) {
+  return (
+    <div className="show-list">
+      <h3>Available Shows</h3>
+      {shows.map((show) => (
+        <Show
+          key={show.id}
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/components/show/ShowList.jsx]: Good key usage. I've seen messaging apps built where they use `key={index}` which will force React to do a lot of work if any of the messages move to another index in the array.

File: podcast-player/podcast-player-app/src/globals.css [hunk 15]
```diff
@@ -0,0 +1,239 @@
+:root {
+}
+body {
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/globals.css]: Not going to dive into this - there's so many ways to style React apps, and very few people use vanilla CSS. Given the current scope of the project vanilla CSS is great (Keep It Simple Stupid) but if you'd like to chat more about alternative styling solutions let me know.

Currently, I'm trying out [Linaria](https://github.com/callstack/linaria), which is very similar to [styled-components](https://styled-components.com/). The big pro behind Linaria is that it's zero-runtime. Styled-components does a lot of work on the client-side to parse the styles into something useable by the browser, while Linaria does all of the heavy lifting at build time.

File: podcast-player/podcast-player-app/src/utils/api-utils.js [hunk 16]
```diff
@@ -0,0 +1,42 @@
+import { removeHtml, parseDate } from "./utils";
+import { parse } from "arraybuffer-xml-parser";
+export async function getShows() {
+  const url = "http://localhost:3500/podcasts";
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/utils/api-utils.js]: Dead code? `localhost` isn't going to work in prod.

File: podcast-player/podcast-player-app/src/utils/utils.js [hunk 17]
```diff
@@ -0,0 +1,17 @@
+export function removeHtml(str) {
+  const regex = new RegExp(/<\/?[^>]+(>|$)/, "ig");
+  const cleaned = str.replaceAll(regex, "");
+  return cleaned;
+}
+
+export function parseDate(str) {
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/utils/utils.js]: Ah - I forgot podcast descriptions can include HTML. Assuming this doesn't introduce any XSS vulnerabilities, it doesn't like unmatched `<` symbols.

`removeHtml('this < that') === 'this '`

Writing your own HTML sanitizer / stripper is incredibly difficult, dangerous, and complex. If you introduce an XSS vulnerability, your whole app is compromised. You also have to worry about handling malformed inputs. RSS XML data is structured, but the description can be an arbitrary string. You can't assume it's going to be well formed.

I'd immediately reject any homebrew solution to validate or sanitize HTML. Some digging found this solution using `DOMParser`: https://stackoverflow.com/a/47140708/1388336

File: podcast-player/podcast-player-app/src/utils/api-utils.js [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+      enclosure: { ...item.enclosure },
+    };
+  });
+
+  const cleanedXML = {
+    channel: {
+      copyright: parsedXml.rss.channel.copyright,
+      description: removeHtml(parsedXml.rss.channel.description),
+      image: { ...parsedXml.rss.channel.image },
+      episodes: [...cleanedEpisodes],
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/utils/api-utils.js]: No need to make a shallow copy of this

File: podcast-player/podcast-player-app/src/components/episode/EpisodeModal.jsx [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+import ReactDOM from "react-dom";
+
+import { ReactComponent as PlayButton } from "../../assets/SVG/play2.svg";
+import { ReactComponent as CloseButton } from "../../assets/SVG/cross.svg";
+
+const Backdrop = (props) => {
+  return <div className="backdrop" onClick={props.onConfirm} />;
+};
+
+const ModalOverlay = ({
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/components/episode/EpisodeModal.jsx]: Interesting that you implemented your own modal. Finding a good component library can be hard.

Modals are notoriously difficult to get right. A good library will handle complex cases out of the box such as keyboard focus trapping and smaller screen sizes (eg: mobile)

File: podcast-player/podcast-player-app/src/components/episode/EpisodePlayer.jsx [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+  // const { $length, $type, $url } = content;
+  return (
+    <div className="player">
+      <audio
+        controls
+        src={episode.enclosure.$url}
```
Comment (matthewoates) [inline podcast-player/podcast-player-app/src/components/episode/EpisodePlayer.jsx]: If we're already massaging the parsed XML data, we should probably also transform the episode data into a more useable shape.

**Review by matthewoates — CHANGES_REQUESTED** [review]

**Comment by matthewoates** [comment]
This is legit. ~~I wasn't able to find any bugs~~. I only found one bug, but it's a big one. There's some UX limitations, and after looking at the code some big technical limitations as well. I'm listening to one of the episodes as I write this code review.

From a UX perspective before looking at any of the code:

- 👍👍 Live demo! This makes review so much easier.
- ~~👍👍 No bugs found!~~ 👎👎 The Apology Line fails to load, and there's no error state to indicate that
- 👍👍 persistent player. I can continue listening to a podcast while I browse
- 👍👍 loading indicators (not pretty, but we're not showing something bogus like "no episodes" while episodes are loading)
- 👍 clean approachable (but not pretty) UI. A Hacker News vibe
- 👎 no visual reflection of selected show or episode within the lists
- 👎👎 no routing / persistence of state via the URL. If you refresh, you're back at the homepage. You can't share a link to an episode with anyone.

And looking at the code, see inline comments + these notes:
- 👎 The folder structure is a little funny. `podcast-player/podcast-player/README.md`. Unnecessarily deep file structures add bloat to projects. The instructions suggest putting everything directly in the `podcast-player` folder.
- 👎👎 RSS feed data is hardcoded in the codebase, but we're also fetching the data on the client-side. Why is this hardcoded?

---

Hopefully you found this to be a valuable exercise. If you're still pumped about this project and would like to take it to the next level, the most notable improvements to make from my perspective are:

- add podcast search (I spent a few minutes looking for a podcast search API but nothing jumped out at me. This also adds a lot of complexity to the project since you'd probably what to make a server. This allows you to hide your API key)
- Use a UI component library. Designing a good looking and functional app from scratch is a pretty big ask and demands a lot of different skills. You can of course use other apps as reference. When it comes to implementation, the React UI libraries I've enjoyed working with are [material-ui](https://github.com/mui/material-ui) and [mantine](https://github.com/mantinedev/mantine).

File: podcast-player/podcast-player-app/README.md (hunk 1 above)
Comment (ramseylove) [reply podcast-player/podcast-player-app/README.md]: changed to js, but then now removed this readme

File: podcast-player/podcast-player-app/data/episodes.js (hunk 2 above)
Comment (ramseylove) [reply podcast-player/podcast-player-app/data/episodes.js]: unused file relating to local xml dummy server

File: podcast-player/podcast-player-app/data/server.js (hunk 3 above)
Comment (ramseylove) [reply podcast-player/podcast-player-app/data/server.js]: deleted with cleanup

File: podcast-player/podcast-player-app/src/utils/api-utils.js (hunk 4 above)
Comment (ramseylove) [reply podcast-player/podcast-player-app/src/utils/api-utils.js]: Now only running parseDate if pubDate exists on show

File: podcast-player/podcast-player-app/src/components/episode/EpisodeDetails.jsx (hunk 6 above)
Comment (ramseylove) [reply podcast-player/podcast-player-app/src/components/episode/EpisodeDetails.jsx]: yup but thinking I might make a different view of the episode info, expanded from modal view but with more data. not sure. do

File: podcast-player/podcast-player-app/src/assets/SVG/circle-right.svg (hunk 5 above)
Comment (ramseylove) [reply podcast-player/podcast-player-app/src/assets/SVG/circle-right.svg]: These are from icomoon.io, the free icons. Where do you recommend putting license info for things used on a site that I did not create? As they do not have any info the code you copy from them.

File: podcast-player/podcast-player-app/src/components/episode/EpisodeList.jsx (hunk 7 above)
Comment (ramseylove) [reply podcast-player/podcast-player-app/src/components/episode/EpisodeList.jsx]: for the selectedShow state, I would need set the show data to the state without using the spread syntax. I get an error with it set to null

File: podcast-player/podcast-player-app/src/assets/SVG/circle-right.svg (hunk 5 above)
Comment (matthewoates) [reply podcast-player/podcast-player-app/src/assets/SVG/circle-right.svg]: The license will say how you should handle the assets. Some are very forgiving (such as [CC0](https://creativecommons.org/share-your-work/public-domain/cc0/) ) which basically means you can do whatever you want and don't need to give credit

File: podcast-player/src/components/episode/EpisodeDetails.jsx [hunk 21]
```diff
@@ -0,0 +1,29 @@
+import React from "react";
+import { ReactComponent as PlayButton } from "../../assets/SVG/play2.svg";
+
+function EpisodeDetails({ episode, setSelectedEpisodePlaying }) {
```
Comment (matthewoates) [inline podcast-player/src/components/episode/EpisodeDetails.jsx:4]: still have this dead code

File: podcast-player/src/components/episode/EpisodeList.jsx [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+    let isCancelled = false;
+
+    if (show && !isCancelled) {
+      setIsLoading(true);
+      getEpisodes(show.feed)
+        .then((data) => {
+          setSelectedShow({ ...data.channel });
+          setEpisodes([...data.channel.episodes]);
+          // temp limiting data
+          setFirstTenEpisodes([...data.channel.episodes].slice(0, 9));
```
Comment (matthewoates) [inline podcast-player/src/components/episode/EpisodeList.jsx:29]: Unnecessary shallow copying, but beyond that -

No need to have this all as separate state hooks. You're not copying from props, but there's a similar issue described here: https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html

You can just save the `data` and use that.

File: podcast-player/src/components/episode/EpisodePlayer.jsx [hunk 23]
```diff
@@ -0,0 +1,17 @@
+import React from "react";
+
+function EpisodePlayer({ episode }) {
+  // const { $length, $type, $url } = content;
```
Comment (matthewoates) [inline podcast-player/src/components/episode/EpisodePlayer.jsx:4]: There's still commented code. Did you do a self-review pass of your own changes?

**Review by matthewoates — CHANGES_REQUESTED** [review]
A few more gripes, and after that this is a pass. Nice work with the quick turnaround!

**Pull request merged**